sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -0,0 +1,1305 @@
+                          (std::is_same<ArgT, sycl::nd_item<3>>::value)>::type
+  runIterationSpace(const sycl::range<3> &LocalSize,
+                    const sycl::range<3> &GlobalSize,
+                    const sycl::id<3> &GlobalOffset) {
+    auto wrappedLambda_NDITEM_3DIM =
+        makeWrapper_NDITEM_3DIM(MKernel, LocalSize, GlobalSize, GlobalOffset);
+
+    spaceDim[0] = (uint32_t)LocalSize[0];
+    spaceDim[1] = (uint32_t)LocalSize[1];
+    spaceDim[2] = (uint32_t)LocalSize[2];
+
+    groupDim[0] = (uint32_t)(GlobalSize[0] / LocalSize[0]);
+    groupDim[1] = (uint32_t)(GlobalSize[1] / LocalSize[1]);
+    groupDim[2] = (uint32_t)(GlobalSize[2] / LocalSize[2]);
+
+    PLACEHOLDER_UNUSED(wrappedLambda_NDITEM_3DIM);
+    // TODO : Invoke invokeLambda_NDITEM_3DIM through CM's multi-threaded
+    // kernel launching with wrappedLambda_NDITEM_3DIM and dimension info
+  }
+};
+
+/// Implementation for ESIMD_CPU device interface accessing ESIMD
+/// intrinsics and LibCM functionalties requred by intrinsics
+struct ESIMDDeviceInterfaceImpl : public ESIMDDeviceInterface {
+  // Function pointer type definitions
+  using BarrierPtr_t = void (*)(void);
+  using SplitBarrierPtr_t = void (*)(uint);
+  using MemoryBasePtr_t = char *(*)(int);
+  using SlmBasePtr_t = char *(*)(void);
+  using SlmSizePtr_t = void (*)(size_t);
+
+  // Function pointers
+
+  // Intrinsics
+  BarrierPtr_t grp_barrier_ptr;
+  SplitBarrierPtr_t split_barrier_ptr;
+  BarrierPtr_t fence_ptr;
+
+  // libcm functionalities used for intrinsics such as
+  // surface/buffer/slm access
+  MemoryBasePtr_t get_surface_base_ptr;
+  SlmBasePtr_t get_slm_base_ptr;
+  SlmSizePtr_t set_slm_size_ptr;
+
+  ESIMDDeviceInterfaceImpl() {
+#ifdef __GNUC__
+    // Linux
+    void *handle = dlopen("libcm.so", RTLD_LAZY);
+#define DLSYM_MAPPING(VAR, TYPE, SYMBOL)                                       \
+  VAR = (TYPE)(dlsym(handle, SYMBOL));                                         \
+  assert(VAR != nullptr);
+
+    // Intrinsics
+    DLSYM_MAPPING(grp_barrier_ptr, BarrierPtr_t, "_Z10cm_barrierv");
+    DLSYM_MAPPING(split_barrier_ptr, SplitBarrierPtr_t, "_Z11cm_sbarrierj");
+    DLSYM_MAPPING(fence_ptr, BarrierPtr_t,
+                  "_Z8cm_fencev"); // With/without bitmask?
+
+    // libcm functionalities used for intrinsics such as surface/slm
+    // access
+    DLSYM_MAPPING(get_surface_base_ptr, MemoryBasePtr_t,
+                  "_Z21get_surface_base_addri");
+    DLSYM_MAPPING(get_slm_base_ptr, SlmBasePtr_t, "_Z16__cm_emu_get_slmv");
+    DLSYM_MAPPING(set_slm_size_ptr, SlmSizePtr_t, "_Z11cm_slm_initj");
+
+#undef DLSYM_MAPPING
+#else
+    // Windows
+    HINSTANCE hcmLib = LoadLibrary(TEXT("libcm.dll"));
+#define PROC_MAPPING(VAR, TYPE, SYMBOL)                                        \
+  VAR = (TYPE)(GetProcAddress(hcmLib, SYMBOL));                                \
+  assert(VAR != nullptr);
+
+    // Intrinsics
+    PROC_MAPPING(grp_barrier_ptr, BarrierPtr_t, "?cm_barrier@@YAXXZ");
```

Comment:
Is it possible to make libcm export unmangled functions(aka  `extern "C"`) to avoid possible problems with how compilers mangle functions?

Reply:
I'm not sure. I would like to push current implementation as it is, and I will try 'unmangled' function names when I bring up CM functionality in later PR.